idrstream/merge_CP_DP.py

```diff
@@ -0,0 +1,162 @@
+    # if batch data have different number of cells, raise an error because they must not have close segmentations
+    if cp_batch_data.shape[0] != dp_batch_data.shape[0]:
+        raise IndexError("Batch data have different number of rows (cells)!")
+
+    # hide warning for pandas chained assignment
+    pd.options.mode.chained_assignment = None
+
+    # get cp and dp column names
+    cp_columns = cp_batch_data.columns
+    dp_columns = dp_batch_data.columns
+    # get metadata columns (columns that show up in both dataframes)
+    metadata_columns = [col for col in cp_columns if col in dp_columns]
+
+    # remove metadata columns from cp and dp columns
+    cp_columns = set(cp_columns) - set(metadata_columns)
+    dp_columns = set(dp_columns) - set(metadata_columns)
+
+    # add CP and DP prefixes to their respective columns
+    cp_batch_data = cp_batch_data.rename(
+        columns={col: f"CP__{col}" for col in cp_columns}
+    )
+    dp_batch_data = dp_batch_data.rename(
+        columns={col: f"DP__{col}" for col in dp_columns}
+    )
+
+    # get each image path because cells within the same image are trying to be associated
+    image_paths = cp_batch_data["Metadata_DNA"].unique()
+
+    # list to compile merged dataframes from each image
+    compiled_merged_data = []
+
+    # iterate through each image to focus on matching cell positions (x,y)
+    for image_path in image_paths:
+        # only work with data from the image of interest
+        cp_image_data = cp_batch_data.loc[cp_batch_data["Metadata_DNA"] == image_path]
+        dp_image_data = dp_batch_data.loc[dp_batch_data["Metadata_DNA"] == image_path]
+
+        # create a location column with x and y coordinates as tuple
+        cp_image_data["Full_Location"] = list(
+            zip(
+                cp_image_data["Location_Center_X"].astype(int),
+                cp_image_data["Location_Center_Y"].astype(int),
+            )
+        )
+        dp_image_data["Full_Location"] = list(
+            zip(
+                dp_image_data["Location_Center_X"].astype(int),
+                dp_image_data["Location_Center_Y"].astype(int),
+            )
+        )
+
+        # make location for dp match the closest cp location (distance minimuzed with hypotenuse)
+        dp_image_data["Full_Location"] = dp_image_data["Full_Location"].map(
+            lambda dp_coord: min(
+                cp_image_data["Full_Location"],
+                key=lambda cp_coord: math.hypot(
+                    cp_coord[0] - dp_coord[0], cp_coord[1] - dp_coord[1]
+                ),
+            )
+        )
+
+        # drop metadata columns from DP before merge
+        dp_image_data = dp_image_data.drop(columns=metadata_columns)
+
+        # merge cp and dp data on location
+        merged_image_data = pd.merge(cp_image_data, dp_image_data, on="Full_Location")
+        # remove of full location column
+        merged_image_data = merged_image_data.drop(columns=["Full_Location"])
```

Comment:
Could the `drop`'s be grouped together to save a line or two (perhaps taking advantage of `list` concatenation)?

Reply:
I don't think so...
We don't want to remove `Full_Location` before merge as this is the column that is merged on.
Leaving the metadata to remove after the merge results in every metadata column being named with a prefix to distinguish the duplicate columns.